lib/queue.js

```diff
@@ -259,104 +500,199 @@ Queue.prototype.finishJob = function (err, data, job, cb) {
+  const jobTick = () => {
+    if (this.paused) {
+      this.queued -= 1;
       return;
     }
 
-    // invariant: in this code path, self.running < self.concurrency, always
-    // after spoolup, self.running + self.queued === self.concurrency
-    self.getNextJob(function (getErr, job) {
-      /* istanbul ignore if */
-      if (getErr) {
-        self.emit('error', getErr);
-        return setImmediate(jobTick);
+    // invariant: in this code path, this.running < this.concurrency, always
+    // after spoolup, this.running + this.queued === this.concurrency
+    this._getNextJob().then((job) => {
+      // We're shutting down.
+      if (this.paused) {
+        // This job will get picked up later as a stalled job if we happen to
+        // get here. We can't easily process this job because at this point
+        // Queue#close has already captured the activeJobs set in a Promise.all
+        // call, and we'd prefer to delay a job than half-process it.
+        this.queued -= 1;
+        return;
       }
 
-      self.running += 1;
-      self.queued -= 1;
-      if (self.running + self.queued < self.concurrency) {
-        self.queued += 1;
+      this.running += 1;
+      this.queued -= 1;
+      if (this.running + this.queued < this.concurrency) {
+        this.queued += 1;
         setImmediate(jobTick);
       }
 
-      self.runJob(job, function (err, status, result) {
-        self.running -= 1;
-        self.queued += 1;
+      if (!job) {
+        // Per comment in Queue#_waitForJob, this branch is possible when the
+        // job is removed before processing can take place, but after being
+        // initially acquired.
+        setImmediate(jobTick);
+        return;
+      }
 
-        /* istanbul ignore if */
-        if (err) {
-          self.emit('error', err);
-        } else {
-          self.emit(status, job, result);
-        }
+      return this._runJob(job).catch((err) => {
+        this.emit('error', err);
+      }).then((results) => {
+        this.running -= 1;
+        this.queued += 1;
 
         setImmediate(jobTick);
+
+        /* istanbul ignore else */
+        if (results) {
+          const status = results[0], result = results[1];
+          this.emit(status, job, result);
+        }
       });
-    });
+    }, (err) => {
+      setImmediate(jobTick);
+      throw err;
+    }).catch((err) => this.emit('error', err));
   };
 
-  var restartProcessing = function () {
-    // maybe need to increment queued here?
-    self.bclient.once('ready', jobTick);
-  };
-  this.bclient.on('error', restartProcessing);
-  this.bclient.on('end', restartProcessing);
+  this.checkStalledJobs(jobTick);
+  this._processDelayed();
 
-  this.checkStalledJobs(setImmediate.bind(null, jobTick));
+  return this;
 };
 
-Queue.prototype.checkStalledJobs = function (interval, cb) {
-  var self = this;
-  cb = typeof interval === 'function' ? interval : cb || helpers.defaultCb;
+Queue.prototype._doStalledJobCheck = function () {
```

Comment:
This separation actually reduced complexity of `_checkStalledJobs` for me - I'd prefer to keep it as-is. They represent distinct functions, so I think it makes sense to leave them as distinct _functions_ :p